src/user/reporting/csv.library.tsx

```diff
@@ -0,0 +1,132 @@
+import tweetnacl from "tweetnacl";
+interface DownloadProps {
+  onComplete?: () => void;
+  onError?: () => void;
+}
+
+export function useDownloadCSV(props: DownloadProps = {}) {
+  const [loading, setLoading] = useState(false);
+  const [error, setError] = useState<string>();
+
+  const download = useCallback(
+    (campaignId: string, isVac: boolean, privateKey?: string) => {
+      setLoading(true);
+      setError(undefined);
+
+      const baseUrl = `/report/campaign/csv/${campaignId}`;
+      fetch(buildAdServerEndpoint(isVac ? `${baseUrl}/vac` : baseUrl), {
+        method: "GET",
+        mode: "cors",
+        credentials: "include",
+        headers: {
+          "Content-Type": "text/csv",
+        },
+      })
+        .then((res) => {
+          if (res.status !== 200) {
+            const err = "Unable to download CSV";
+            setError(err);
+            throw new Error(err);
+          }
+
+          return res.blob();
+        })
+        .then((blob) => {
+          const file = new Blob([blob], {
+            type: "text/csv",
+            endings: "transparent",
+          });
+
+          if (isVac) {
+            return transformConversionEnvelope(file, privateKey);
+          }
+
+          return Promise.resolve(file);
+        })
+        .then((file) => {
+          const fileURL = URL.createObjectURL(file);
+          const link = document.createElement("a");
+          link.href = fileURL;
+          link.setAttribute("download", `${campaignId}.csv`);
+          document.body.appendChild(link);
+          link.click();
+          if (props.onComplete) props.onComplete();
+        })
+        .catch((e) => {
+          setError(e.message);
+        })
+        .finally(() => {
+          setLoading(false);
+        });
+    },
+    [],
+  );
+
+  return { download, loading, error };
+}
+
+type Envelope = { ciphertext: string; epk: string; nonce: string };
+async function transformConversionEnvelope(
+  blob: Blob,
+  privateKey?: string,
+): Promise<Blob> {
+  if (!privateKey) {
+    return Promise.resolve(blob);
+  }
+
+  const te = new TextEncoder();
+  const td = new TextDecoder();
+  const ui8a = (s: string) => Uint8Array.from(atob(s), (c) => c.charCodeAt(0));
+
+  return await new Promise((resolve, reject) => {
+    const fileReader = new FileReader();
+    fileReader.onload = () => {
+      const text = fileReader.result as string | null;
+      if (text === null) {
+        reject(new Error("No file Result"));
+        return;
+      }
+
+      try {
+        Papa.parse(text, {
+          header: true,
+          escapeChar: "\\",
+          transform(value: string, field: string) {
+            if (field.includes("Conversion")) {
+              const { ciphertext, nonce, epk }: Envelope = JSON.parse(value);
+              const res = tweetnacl.box.open(
+                ui8a(ciphertext),
+                ui8a(nonce),
+                ui8a(epk),
+                ui8a(privateKey),
```

Comment:
As mentioned on the review page, the privateKey array could be zeroed after use rather than having tons of copies of it in memory. However, the array is being created from a string anyway, so there's not a huge benefit.